src/text_shaper/directwrite_locator.cpp

```diff
@@ -0,0 +1,227 @@
+        ComPtr<IDWriteFontFallback> systemFontFallback;
+
+        std::wstring userLocale;
+
+        Private()
+        {
+            auto hr = DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED,
+                __uuidof(IDWriteFactory7),
+                reinterpret_cast<IUnknown**>(factory.GetAddressOf()));
+
+            wchar_t locale[LOCALE_NAME_MAX_LENGTH];
+            GetUserDefaultLocaleName(locale, sizeof(locale));
+            userLocale = locale;
+
+            factory->GetSystemFontCollection(&systemFontCollection);
+            factory->GetSystemFontFallback(&systemFontFallback);
+        }
+    };
+
+    directwrite_locator::directwrite_locator() :
+        d{ new Private(), [](Private* p) { delete p; } }
+    {
+    }
+
+    directwrite_locator::~directwrite_locator()
+    {
+    }
+
+    font_source_list directwrite_locator::locate(font_description const& _fd)
+    {
+        font_source_list output;
+
+        // TODO: use libunicode for that (TODO: create wchar_t/char16_t converters in libunicode)
+        std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> wStringConverter;
+        std::wstring familyName = wStringConverter.from_bytes(_fd.familyName);
+
+        UINT32 familyIndex;
+        BOOL familyExists = FALSE;
+        d->systemFontCollection->FindFamilyName(familyName.data(), &familyIndex, &familyExists);
+
+        if (!familyExists) {
+            // Fallback to Consolas
+            const wchar_t* consolas = L"Consolas";
+            d->systemFontCollection->FindFamilyName(consolas, &familyIndex, &familyExists);
+        }
+
+        ComPtr<IDWriteFontFamily> fontFamily;
+        d->systemFontCollection->GetFontFamily(familyIndex, &fontFamily);
+
+        for (UINT32 k = 0, ke = fontFamily->GetFontCount(); k < ke; ++k)
+        {
+            ComPtr<IDWriteFont> font;
+            fontFamily->GetFont(k, font.GetAddressOf());
+
+            font_weight weight = dwFontWeight(font->GetWeight());
+            if (weight != _fd.weight)
+                continue;
+
+            font_slant slant = dwFontSlant(font->GetStyle());
+            if (weight != _fd.weight)
+                continue;
+
+            ComPtr<IDWriteFontFace> fontFace;
+            font->CreateFontFace(&fontFace);
+            output.emplace_back(font_path{ wStringConverter.to_bytes(get_font_path(fontFace.Get())) });
+        }
+
+        return output;
+    }
+
+    font_source_list directwrite_locator::all()
+    {
+        // TODO;
+        return {};
+    }
+
+    font_source_list directwrite_locator::resolve(gsl::span<const char32_t> codepoints)
+    {
+        font_source_list output;
+
+        std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> conv1;
+        std::string bytes = conv1.to_bytes(std::u32string{ codepoints.data(), codepoints.size() });
+        std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> conv2;
+        std::wstring wText = conv2.from_bytes(bytes);
```

Comment:
um, it's converting from UTF-32 to UTF-8 to UTF-16? haha. That is hilarious!

No need to change that in this PR, as mentioned above, I'll fix libunicode first so we can do it in *one* pass.